Fix GHC bootstrap in pkgsMusl and include patch for binutils/16177

expipiplus1 opened the pull request:
###### Motivation for this change

This allows one to cross compile statically linked Haskell programs for armv7

Fixes https://github.com/NixOS/nixpkgs/issues/85924 using Ben's patch from there. A similar patch should probably be applied to ghc8102-binary.

Fixes https://sourceware.org/bugzilla/show_bug.cgi?id=16177 by applying the patch mentioned there.

Although there was some talk about the efficacy of the binutils patch (https://sourceware.org/bugzilla/show_bug.cgi?id=16177#c9 and expipiplus1@14efdb2) the resulting binary seems to run without issue on the target platform. Jessica's patch there caused ld to fail linking some programs.

It would be good if there was some resolution to the discussion on the binutils ticket, however this does seem to be an improvement over the broken state of things at the moment. 

Obviously this would require rebuilding everything Haskell, and everything on armv7.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

expipiplus1 committed:
ghc865-binary: Use binary distribution which links against ncurses6 for
x86_64-linux

Ben Gamari's patch from #85924.

Fixes #85924, allowing one to bootstrap GHC in `pkgsMusl`

`nix-build -A pkgsMusl.haskellPackages.hello` succeeds with this patch.

expipiplus1 committed:
binutils: patch to fix https://sourceware.org/bugzilla/show_bug.cgi?id=16177

This bug was preventing one compiling Haskell programs from `pkgsMusl` for
armv7.

`nix-build --argstr crossSystem "armv7l-linux" -A pkgsMusl.haskellPackages.hello`
succeeds with this patch.

The patch is Nick Clifton's one, rebased by @ericson2314 here
https://sourceware.org/bugzilla/show_bug.cgi?id=16177#c6

Although there was some talk about the efficacy of the binutils patch
(https://sourceware.org/bugzilla/show_bug.cgi?id=16177#c9) the resulting
binary seems to run without issue on the target platform. Jessica's
patch there caused ld to fail linking some programs. Nick's proposed
patch has worked well in my testing so far (a Haskell project of some
small complexity cross compiled with musl to armv7).

expipiplus1 commented:
@ofborg build pkgsMusl.haskellPackages.hello

cdepillabout commented:
cc @ericson2314 as maintainer for binutils

dasJ commented:
Mass rebuild so open this against staging, please

expipiplus1 commented:
Oops, changed the target branch in github and the whole world got requested for reviews, sorry!

dasJ commented:
@expipiplus1 Don't worry about it, that's just what happens :shrug:

expipiplus1 commented:
I've been using this for a while with no problems, no issues with the binaries on the host machine either.

cdepillabout commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -79,7 +79,8 @@ stdenv.mkDerivation {
     ./0001-x86-Properly-add-X86_ISA_1_NEEDED-property.patch
-  ] ++ lib.optional stdenv.targetPlatform.isiOS ./support-ios.patch;
+  ] ++ lib.optional stdenv.targetPlatform.isiOS ./support-ios.patch
+    ++ lib.optional stdenv.targetPlatform.isAarch32 ./R_ARM_COPY.patch;
```
Could you leave a short comment above this line with a description of what this patch does, why it is needed, and when we can expect to remove it (if ever)?

cdepillabout reviewed (changes requested):
No one has seemed to speak out against this patch, so I'd like to merge it as soon as the the comment below has been added.

cdepillabout commented:
@expipiplus1 Oh, also, I couldn't quite understand from your PR message, but what are a few specific derivations that this PR should get building (that were not buildable before).  It should only be armv7/musl/haskell-related stuff, right?

expipiplus1 commented:
Thanks @cdepillabout, I've expanded the commit messages and added a comment on the suggested line.

I rebased on staging and am just testing things now.

expipiplus1 commented:
@nh2 I think this should interest you

Ericson2314 reviewed (approved):
Looks great to me!

cdepillabout commented:
@expipiplus1 After looking at this a little more, I don't quite understand what is going on.

The PR message specifically says that this "allows one to cross compile statically linked Haskell programs for armv7".  This PR includes patches for `8.6.5-binary.nix`.  However, when building ghc884, it looks like `ghc8102BinaryMinimal` is used if `isArch64`.

When _cross compiling_ for armv7, I'm guessing that `ghc865Binary` is used instead of `ghc8102BinaryMinimal`?

Also, you may be aware that `haskell-updates` has moved to track stackage nightly (https://github.com/NixOS/nixpkgs/pull/103740), which means that the Haskell package set will be built with ghc8102 (although this shouldn't directly affect anything, since `ghc8102` has the same logic around it's `bootPkgs` Haskell compiler).  However, when stackage moves to ghc9, it looks like `bootPkgs` switches to `ghc8102Binary`, so it might be possible that ghc8102Binary will need to be patched sooner rather than later.

------------------------------------------------------------

@GrahamcOfBorg pkgsMusl.haskellPackages.lens

expipiplus1 commented:
> The PR message specifically says that this "allows one to cross compile statically linked Haskell programs for armv7". This PR includes patches for `8.6.5-binary.nix`. However, when building ghc884, it looks like `ghc8102BinaryMinimal` is used if `isArch64`.
> 
> When _cross compiling_ for armv7, I'm guessing that `ghc865Binary` is used instead of `ghc8102BinaryMinimal`?

Yes, because `armv7l` is a 32 bit arm platform :D

> Also, you may be aware that `haskell-updates` has moved to track stackage nightly (#103740), which means that the Haskell package set will be built with ghc8102 (although this shouldn't directly affect anything, since `ghc8102` has the same logic around it's `bootPkgs` Haskell compiler). However, when stackage moves to ghc9, it looks like `bootPkgs` switches to `ghc8102Binary`, so it might be possible that ghc8102Binary will need to be patched sooner rather than later.

I wonder if this is already fixed for 8102, it's using the deb10 distribution and linking against ncurses 6 already without replacing libtinfo.so, i.e. the equivalent patch as this one for 865 has already been applied.

expipiplus1 commented:
Since opening this PR `pkgsMusl.gdb` has bitrotted so `pkgsMusl.haskellPackages.hello` no longer compiles (why does it depend on gdb anyway...)

Ericson2314 commented:
> Since opening this PR `pkgsMusl.gdb` has bitrotted so `pkgsMusl.haskellPackages.hello` no longer compiles (why does it depend on gdb anyway...)

I bet the problem is that python has a bad dependency on gdb, such that build-time python (e.g. used in Sphinx) wants a cross gdb.

expipiplus1 commented:
`nb --argstr crossSystem "armv7l-linux" -A pkgsMusl.haskell.packages.ghc901.hello` fails building `ghc901` so I can't test (although it's outside the scope of this PR)

```
/nix/store/79rshjqnibwcx468m4i31wnmasaj6nf4-binutils-2.31.1/bin/ld: /build/ghc-9.0.0.20200925/compiler/stage1/build/libHSghc-9.0.0.20200925.a(Plugins.o): in function `ghc_GHCziDriverziPlugins_keepRenamedSource_info':
(.text+0x245c): undefined reference to `ghc_GHCziCoreziOptziMonad_zdfxMonadCoreM_closure'
/nix/store/79rshjqnibwcx468m4i31wnmasaj6nf4-binutils-2.31.1/bin/ld: /build/ghc-9.0.0.20200925/compiler/stage1/build/libHSghc-9.0.0.20200925.a(Plugins.o): in function `ghc_GHCziDriverziPlugins_defaultPluginzux_info':
(.text+0x2494): undefined reference to `ghc_GHCziCoreziOptziMonad_zdfxMonadCoreM_closure'
collect2: error: ld returned 1 exit status
`gcc' failed in phase `Linker'. (Exit code: 1)
make[1]: *** [ghc/ghc.mk:124: ghc/stage1/build/tmp/ghc-stage1] Error 1
make: *** [Makefile:126: all] Error 2
builder for '/nix/store/d2pa11x0pc80wlqdaxaxk2g6n6kd4lbl-ghc-9.0.0.20200925.drv' failed with exit code 2
```

Ericson2314 commented:
Indeed it is. Cython being used + gdb being target sensative + no splicing black magic has gotten here. I'll make an issue.

expipiplus1 commented:
If I revert the gdb change then `pkgsMusl.haskellPackages.hello` builds fine, obviously this update shouldn't be reverted but it's nice to know that this change should enable `pkgsMusl.haskellPackages.hello` to work once gdb is working again.

nh2 commented on pkgs/development/compilers/ghc/8.6.5-binary.nix:
```diff
@@ -38,8 +38,8 @@ stdenv.mkDerivation rec {
     x86_64-linux = {
-      url = "http://haskell.org/ghc/dist/${version}/ghc-${version}-x86_64-deb9-linux.tar.xz";
-      sha256 = "1pqlx6rdjs2110g0y1i9f8x18lmdizibjqd15f5xahcz39hgaxdw";
+      url = "http://haskell.org/ghc/dist/${version}/ghc-${version}-x86_64-fedora27-linux.tar.xz";
```
Please add a comment why we use this one, just to prevent that somebody changes it back to `deb9` to fix some other issue (perhaps also linking https://github.com/NixOS/nixpkgs/issues/85924).

nh2 commented on pkgs/development/tools/misc/binutils/default.nix:
```diff
@@ -101,6 +101,10 @@ stdenv.mkDerivation {
   ++ lib.optional stdenv.targetPlatform.isiOS ./support-ios.patch
+  ++ # This patch was suggested by Nick Clifton to fix
+     # https://sourceware.org/bugzilla/show_bug.cgi?id=16177
+     # It can be removed when that 7-year-old bug is closed.
```
Please add to the comment what the patch's _purpose_ is / why nixpkgs needs it; it's not immediately obvious even from the issue description. Linking to https://github.com/NixOS/nixpkgs/issues/85924 would also be fine.

nh2 reviewed (approved):
I added two requests to add comments, sounds great otherwise.

expipiplus1 commented:
@nh2 I've added the comments as suggested

Also, while doing this I realize that this patch will break the i686 build! I've added some small amount of complexity to only use ncurses6 on x86_64.

nh2 reviewed (approved):
> Also, while doing this I realize that this patch will break the i686 build! I've added some small amount of complexity to only use ncurses6 on x86_64.

Just to understand, why is that? Does `ncurses6` not work on i686?

expipiplus1 commented:
It doesn't compile with the changes as before for i686, I guess because it's still using the deb9 version which is linked with ncurses5. There's no fedora27 version for i686.

nh2 commented:
> It doesn't compile with the changes as before for i686, I guess because it's still using the deb9 version which is linked with ncurses5. There's no fedora27 version for i686.

@expipiplus1 Ah, please add that as a comment as well -- "upstream does not provide a version" is a good justification.

@bgamari Do you think there _should_ be a fedora27 version for i686, or should this stay with ncurses5?

expipiplus1 commented:
Is there even an i686 release of fedora27?

expipiplus1 commented:
@nh2, done

nh2 commented:
> Is there even an i686 release of fedora27?

I think they dropped kernel support with fedora31, but I'm not sure I get the whole picture there: https://fedoramagazine.org/in-fedora-31-32-bit-i686-is-86ed/

nh2 commented:
@cdepillabout Good to go? The comment you requested has been added. I'm not quite sure if all your questsions from https://github.com/NixOS/nixpkgs/pull/103183#issuecomment-729393383 are addressed.

cdepillabout reviewed (approved):
This looks good, thanks for fleshing out the comments!

talw commented:
@expipiplus1 @Ericson2314 

> Indeed it is. Cython being used + gdb being target sensative + no splicing black magic has gotten here. I'll make an issue: #104135.

> If I revert the gdb change then pkgsMusl.haskellPackages.hello builds fine, obviously this update shouldn't be reverted but it's nice to know that this change should enable pkgsMusl.haskellPackages.hello to work once gdb is working again.

First of all, thank you for the PRs!

I still seem to fail compiling statically due to gdb, whether with the earliest revision containing both PRs `15b3bc33064eeb0cba743cad585c829b6694669c` or with the latest revision of `nixos-unstable`

```sh
Configuring in ./gdb
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../opcodes -I. -I../../opcodes -I../bfd -I../../opcodes/../include -I../../opcodes/../bfd -I./../intl -W -Wall -Wstrict-prototypes -Wmissing-prototypes -Wshadow -Wstack-usage=262144 -g -O2 -MT disassemble.lo -MD -MP -MF .deps/disassemble.Tpo -c -DARCH_all ../../opcodes/disassemble.c -o disassemble.o
mkdir -p -- ./libctf
Configuring in ./libctf
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: ax-ipa.o: in function `ax_printf(unsigned long, unsigned long, char const*, int, unsigned long*) [clone .isra.0]':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:928: undefined reference to `libintl_gettext'
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: /build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:837: undefined reference to `libintl_gettext'
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: /build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:889: undefined reference to `libintl_gettext'
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: ax-ipa.o: in function `gdb_eval_agent_expr(eval_agent_expr_context*, agent_expr*, unsigned long*)':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:1334: undefined reference to `libintl_gettext'
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: gdbsupport/common-utils-ipa.o: in function `xstrvprintf(char const*, __va_list_tag*)':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/../gdbsupport/common-utils.cc:57: undefined reference to `libintl_gettext'
/nix/store/nz8dyqljf2xwacml7jivafpqqajh15jj-binutils-2.31.1/bin/ld: gdbsupport/common-utils-ipa.o:/build/gdb-10.1/_build/gdbserver/../../gdbserver/../gdbsupport/common-utils.cc:69: more undefined references to `libintl_gettext' follow
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:389: libinproctrace.so] Error 1

(...non-error lines...)

make[2]: Leaving directory '/build/gdb-10.1/_build/gdbserver'
make[1]: *** [Makefile:9365: all-gdbserver] Error 2

(...non-error lines...)

make: *** [Makefile:854: all] Error 2
builder for '/nix/store/lay3ilkzal9qsr2k9nbyxm38x1p5cd8k-gdb-10.1.drv' failed with exit code 2
building '/nix/store/v66vfzmg640vgljbq3w61xsyjn1blaqp-cmake-3.19.1.drv'...
cannot build derivation '/nix/store/cbi55fccyxivc6g1nrkh029r4yzdjwfx-python3.8-Cython-0.29.21.drv': 1 dependencies couldn't be built
```

Is it the same problem? Or is it a different problem?

Ericson2314 commented:
@talw that looks different, can you make a separate issue if this still exists?

Ericson2314 commented:
@matthewbauer should we backport this?

talw commented:
@Ericson2314

I see. I guess it is something different.
I was pulled off temporarily from working on the project where I stumbled upon this case.
But once I get back to it, I will try again with the latest revision at that time, and if it will still happen I shall make an issue for it.

Thanks!

talw commented:
> @talw that looks different, can you make a separate issue if this still exists?

@Ericson2314 

I looked at it today, and it looks like it is different as gdb's derivations are different.
Comparing them with `nix-diff` there were changes going all the way back to `curl`, `zlib` and friends. I am using [static-haskell-nix](https://github.com/nh2/static-haskell-nix) which has a `survey.nix` that overrides a bunch of stuff that results in a `gdb` that fails to compile. My issue is to be in that repository. Unless I am misguided about the usage of `survey.nix`

nh2 commented:
> I am using `static-haskell-nix` which has a `survey.nix` that overrides a bunch of stuff that results in a gdb that fails to compile.

@talw If you can tell me how to reproduce (best even directly with a `nix-build` invocation to build `gdb` from it), I can try and help fix it.

talw commented:
> > I am using `static-haskell-nix` which has a `survey.nix` that overrides a bunch of stuff that results in a gdb that fails to compile.
> 
> @talw If you can tell me how to reproduce (best even directly with a `nix-build` invocation to build `gdb` from it), I can try and help fix it.

Thanks @nh2

For reference here is the message I wrote and deleted shortly after (to which you replied):

>> @talw that looks different, can you make a separate issue if this still exists?

> @Ericson2314 
> 
> I looked at it today, and it looks like it is different as gdb's derivations are different.
> Comparing them with `nix-diff` there were changes going all the way back to `curl`, `zlib` and friends. I am using [static-haskell-nix](https://github.com/nh2/static-haskell-nix) which has a `survey.nix` that overrides a bunch of stuff that results in a `gdb` that fails to compile. My issue is to be in that repository. Unless I am misguided about the usage of `survey.nix`

What I realized shortly after is that while the gdb is different and fails to compile, even without
`survey.nix`'s overrides the vanilla `pkgsMusl.gdb` fails to compile due to the same problem and on the
revision that includes the 2 PRs that I imagined would fix the `gdb` issue (perhaps they are unrelated after all?).

The PRs are https://github.com/NixOS/nixpkgs/pull/104201 and this one https://github.com/NixOS/nixpkgs/pull/103183

The revision I chose is simply the earliest one to contain both PRs. To reproduce:

```sh
$ nix-build --show-trace --no-out-link -E "(import (fetchTarball \"https://github.com/NixOS/nixpkgs/archive/15b3bc33064eeb0cba743cad585c829b6694669c.tar.gz\") {}).pkgsMusl.gdb"

...abbreviated...

/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: ax-ipa.o: in function `ax_printf(unsigned long, unsigned long, char const*, int, unsigned long*) [clone .isra.0]':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:928: undefined reference to `libintl_gettext'
/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: /build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:837: undefined reference to `libintl_gettext'
/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: /build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:889: undefined reference to `libintl_gettext'
/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: ax-ipa.o: in function `gdb_eval_agent_expr(eval_agent_expr_context*, agent_expr*, unsigned long*)':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/ax.cc:1334: undefined reference to `libintl_gettext'
/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: gdbsupport/common-utils-ipa.o: in function `xstrvprintf(char const*, __va_list_tag*)':
/build/gdb-10.1/_build/gdbserver/../../gdbserver/../gdbsupport/common-utils.cc:57: undefined reference to `libintl_gettext'
/nix/store/zkpjbbicsqa36sg6jb7fcg5qlahcx2na-binutils-2.31.1/bin/ld: gdbsupport/common-utils-ipa.o:/build/gdb-10.1/_build/gdbserver/../../gdbserver/../gdbsupport/common-utils.cc:69: more undefined references to `libintl_gettext' follow
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:389: libinproctrace.so] Error 1

...abbreviated...

make[2]: Leaving directory '/build/gdb-10.1/_build/gdbserver'
make[1]: *** [Makefile:9365: all-gdbserver] Error 2

...abbreviated...

make: *** [Makefile:854: all] Error 2
builder for '/nix/store/9ly1l78ggcd6nlhs1c81ilw6x83zc2py-gdb-10.1.drv' failed with exit code 2
error: build of '/nix/store/9ly1l78ggcd6nlhs1c81ilw6x83zc2py-gdb-10.1.drv' failed
```
